### PR TITLE
[core] Preserve friendly name and tensor names in PPP

### PR DESCRIPTION
**Pull request by praasz**
### Details:
- For models with  `version > 10` the node's friendly name and tensor's names will be not moved from original node. If conversion node added by PPP or convert precision then new friendly name will be created based on previous node. The new node name will have format `[previous_node_name].[port_number]`.
- Align `ConvertPrecision` transformation to with PPP
- The issue with lost tensor names when set them on model's inputss/outputs  has been solved in #25954 
- For model version 10 the old behavior is preserved as legacy compatibility mode

### Tickets:
 - [CVS-127482](https://jira.devtools.intel.com/browse/CVS-127482)

### Depends on:
- #25954 - implements tensor names handling for model outputs (should be used here)

### CODE REVIEW

**Comment by mlukasze** [comment]
@itikhono, @KodiaqQ, @alexsu52
this fix awaits for review for a very long time, please take a look.

File: src/core/src/node_output.cpp [hunk 1]
```diff
@@ -97,35 +97,39 @@ void Output<Node>::replace(const Output<Node>& replacement) {
 }
 
 RTMap& Output<Node>::get_rt_info() {
-    return m_node->m_outputs.at(m_index).get_rt_info();
+    return get_tensor().get_rt_info();
+    // return m_node->m_outputs.at(m_index)->get_rt_info();
```
Comment (barnasm1) [inline src/core/src/node_output.cpp:101]: Is that comment needed?

File: src/core/src/node_output.cpp [hunk 2]
```diff
@@ -201,7 +205,7 @@ const PartialShape& Output<const Node>::get_partial_shape() const {
 std::set<Input<Node>> Output<const Node>::get_target_inputs() const {
     std::set<Input<Node>> result;
 
-    for (auto& input : m_node->m_outputs.at(m_index).get_inputs()) {
+    for (auto& input : m_node->m_outputs.at(m_index)->get_inputs()) {
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/src/node_output.cpp:208]: The `input` is const becuase `get_input` returns const vector. But will update it.

File: src/core/src/node_output.cpp (hunk 1 above)
Comment (praasz) [reply src/core/src/node_output.cpp:101]: Removed.

File: src/bindings/python/src/pyopenvino/graph/preprocess/pre_post_process.cpp [hunk 3]
```diff
@@ -370,6 +370,18 @@ static void regclass_graph_OutputTensorInfo(py::module m) {
...
         )");
+
+    info.def(
+        "_set_names_compatibility_mode",
+        [](ov::preprocess::OutputTensorInfo& self, const bool compatibility_mode) {
+            return &self.set_names_compatibility_mode(compatibility_mode);
+        },
+        py::arg("compatibility_mode"),
+        R"(
+            Set keep names compatibility mode
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/preprocess/pre_post_process.cpp:381]:
```suggestion
            Set keep names compatibility mode
 
```
Needs for proper documentation parsing

File: src/bindings/python/src/pyopenvino/graph/preprocess/pre_post_process.cpp [hunk 4]
```diff
@@ -366,10 +366,23 @@ static void regclass_graph_OutputTensorInfo(py::module m) {
...
+
+            :param compatibility_mode: Mode to be set: True enable compatiblity, False disable
+            :type layout: Boolean
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/preprocess/pre_post_process.cpp:384]:
```suggestion
            :type compatibility_mode: bool
```

File: src/bindings/python/src/pyopenvino/graph/preprocess/pre_post_process.cpp [hunk 5]
```diff
@@ -366,10 +366,23 @@ static void regclass_graph_OutputTensorInfo(py::module m) {
         },
         py::arg("layout"),
         R"(
-            Set layout for output tensor info 
+            Set layout for output tensor info
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/preprocess/pre_post_process.cpp:369]:
```suggestion
            Set layout for output tensor info

```

File: src/bindings/python/src/pyopenvino/graph/preprocess/pre_post_process.cpp [hunk 6]
```diff
@@ -419,7 +433,7 @@ static void regclass_graph_OutputModelInfo(py::module m) {
         },
         py::arg("layout"),
         R"(
-            Set layout for output model info 
+            Set layout for output model info
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/preprocess/pre_post_process.cpp:436]:
```suggestion
            Set layout for output model info

```

**Review by KodiaqQ — APPROVED** [review]

File: src/bindings/python/src/pyopenvino/graph/preprocess/pre_post_process.cpp [hunk 7]
```diff
@@ -252,7 +252,7 @@ static void regclass_graph_InputTensorInfo(py::module m) {
         },
         py::arg("layout"),
         R"(
-            Set layout for input tensor info 
+            Set layout for input tensor info
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/preprocess/pre_post_process.cpp:255]:
```suggestion
            Set layout for input tensor info

```

**Review by akuporos — APPROVED** [review]

File: src/core/src/op/result.cpp [hunk 8]
```diff
@@ -9,12 +9,72 @@
...
+          m_output_names{} {};
+
+    void set_names(const std::unordered_set<std::string>& names) override {
+        rm_tensor_output_names();
+        m_output_names = names;
+        get_tensor().add_names(m_output_names);
+    }
+
+    void add_names(const std::unordered_set<std::string>& names) override {
+        m_output_names.insert(names.begin(), names.end());
```
Comment (barnasm1) [inline src/core/src/op/result.cpp:35]: Iteration over `std::unordered_set` might be slow.

File: src/core/tests/preprocess.cpp [hunk 9]
```diff
@@ -1548,9 +1547,36 @@ TEST(pre_post_process, postprocess_convert_element_type_explicit) {
...
-    EXPECT_EQ(names_count, 2);  // last node + result referencing to it
+    EXPECT_EQ(names_count, 2);  // result + node connected to it has same name referencing to it
     EXPECT_EQ(name, f->output().get_node_shared_ptr()->get_friendly_name());
-    EXPECT_EQ(name_last_op,
+    EXPECT_EQ(name_last_op + ".0",
+              f->get_results().front()->get_input_source_output(0).get_node_shared_ptr()->get_friendly_name());
+}
+
+TEST(pre_post_process, trivial_model_convert_element_type_explicit) {
+    auto f = create_trivial(element::f32, Shape{1, 3, 2, 2});
```
Comment (barnasm1) [inline src/core/tests/preprocess.cpp:1557]: Use `const auto` where possible

File: src/core/src/preprocess/preprocess_impls.cpp [hunk 10]
```diff
@@ -369,63 +369,54 @@ void OutputInfo::OutputInfoImpl::build(ov::ResultVector& results) {
...
+        } else if (node.get_node_shared_ptr() != orig_parent) {
+            // Result node is changed - add ".<idx>" suffix
+            node.get_node_shared_ptr()->set_friendly_name(
+                orig_parent->get_friendly_name() + "." +
+                std::to_string(result->get_input_source_output(0).get_index()));
+        }
+        result->input(0).replace_source_output(node);
+        result->revalidate_and_infer_types();
     } else if (node.get_node_shared_ptr() != orig_parent) {
+        // const auto idx = result->get_input_source_output(0).get_index();
```
[the remark this answers is not in the document]
Comment (praasz) [reply src/core/src/preprocess/preprocess_impls.cpp]: Removed.

**Comment by mlukasze** [comment]
@itikhono we want to merge today, please take a last look is everything ok from your perspective

**Review by t-jankowski — APPROVED** [review]
Lgtm as of Core and IE.

File: src/core/tests/preprocess.cpp (hunk 9 above)
Comment (praasz) [reply src/core/tests/preprocess.cpp:1563]: `f` is re-used in the test

File: src/core/src/op/result.cpp (hunk 8 above)
Comment (praasz) [reply src/core/src/op/result.cpp]: After some offline discussion and test it looks is there should not any performance penalty.

File: src/core/tests/preprocess.cpp (hunk 9 above)
Comment (praasz) [reply src/core/tests/preprocess.cpp]: Added in new test only.

**Pull request merged**